tutorial/example-hooks.cc

```diff
@@ -0,0 +1,130 @@
+  fprintf(f, "%p\n", __builtin_return_address(0));
+}
+
+void __sanitizer_cov_trace_cmp8(uint64_t Arg1, uint64_t Arg2) {
+  fprintf(f, "%p: cmp8(%lu, %lu)\n", __builtin_return_address(0), Arg1, Arg2);
+}
+
+void __sanitizer_cov_trace_const_cmp8(uint64_t Arg1, uint64_t Arg2) {
+  fprintf(f, "%p: ccmp8(%lu, %lu)\n", __builtin_return_address(0), Arg1, Arg2);
+}
+
+void __sanitizer_cov_trace_cmp4(uint32_t Arg1, uint32_t Arg2) {
+  fprintf(f, "%p: cmp4(%u, %u)\n", __builtin_return_address(0), Arg1, Arg2);
+}
+
+void __sanitizer_cov_trace_const_cmp4(uint32_t Arg1, uint32_t Arg2) {
+  fprintf(f, "%p: ccmp4(%u, %u)\n", __builtin_return_address(0), Arg1, Arg2);
+}
+
+void __sanitizer_cov_trace_cmp2(uint16_t Arg1, uint16_t Arg2) {
+  fprintf(f, "%p: cmp2(%u, %u)\n", __builtin_return_address(0), Arg1, Arg2);
+}
+
+void __sanitizer_cov_trace_const_cmp2(uint16_t Arg1, uint16_t Arg2) {
+  fprintf(f, "%p: ccmp2(%u, %u)\n", __builtin_return_address(0), Arg1, Arg2);
+}
+
+void __sanitizer_cov_trace_cmp1(uint8_t Arg1, uint8_t Arg2) {
+  fprintf(f, "%p: cmp1(%u, %u)\n", __builtin_return_address(0), Arg1, Arg2);
+}
+
+void __sanitizer_cov_trace_const_cmp1(uint8_t Arg1, uint8_t Arg2) {
+  fprintf(f, "%p: ccmp1(%u, %u)\n", __builtin_return_address(0), Arg1, Arg2);
+}
+
+void __sanitizer_cov_trace_switch(uint64_t Val, uint64_t *Cases) {
+  // TODO: Print which case is taken.
+  fprintf(f, "%p: switch(%lu)\n", __builtin_return_address(0), Val);
+}
+
+void __sanitizer_cov_trace_div4(uint32_t Val) {
+  fprintf(f, "%p: div4(X, %u)\n", __builtin_return_address(0), Val);
+}
+
+void __sanitizer_cov_trace_div8(uint64_t Val) {
+  fprintf(f, "%p: div8(X, %lu)\n", __builtin_return_address(0), Val);
+}
+
+void __sanitizer_cov_trace_gep(uintptr_t Idx) {
+  fprintf(f, "%p: gep(%lu)\n", __builtin_return_address(0), Idx);
+}
+
+void __sanitizer_weak_hook_memcmp(void *caller_pc, const void *s1,
+                                  const void *s2, size_t n, int result) {
+  fprintf(f, "%p: memcmp(%p, %p, %lu) -> %u\n", caller_pc, s1, s2, n, result);
+}
+
+void __sanitizer_weak_hook_strncmp(void *caller_pc, const char *s1,
+                                   const char *s2, size_t n, int result) {
+  fprintf(f, "%p: strncmp(%p, %p, %lu) -> %u\n", caller_pc, s1, s2, n, result);
+}
+
+void __sanitizer_weak_hook_strcmp(void *caller_pc, const char *s1,
+                                   const char *s2, int result) {
+  fprintf(f, "%p: strcmp(%p, %p) -> %u\n", caller_pc, s1, s2, result);
+}
+
+void __sanitizer_weak_hook_strncasecmp(void *caller_pc, const char *s1,
+                                       const char *s2, size_t n, int result) {
+  fprintf(f, "%p: strncasecmp(%p, %p, %lu) -> %u\n", caller_pc, s1, s2, n,
+          result);
+}
+
+void __sanitizer_weak_hook_strcasecmp(void *caller_pc, const char *s1,
+                                      const char *s2, int result) {
+  fprintf(f, "%p: strcasecmp(%p, %p) -> %u\n", caller_pc, s1, s2, result);
```

Comment:
I hope we are not going  to parse this format, or we will have trouble to distinguish "," from string and from decorators

Reply:
For these examples, I'm just printing the address of the pointer, not the string itself.  The goal was more of a PoC rather than a final product.